In [75]:
import pandas as pd
from pathlib import Path

First, read in the imdb data and the giantbomb data as dataframes and get a count

In [76]:
imdb_games = pd.read_csv("imdb_games_db.csv", index_col = 0)
giantbomb_games = pd.read_csv("clean_giantbomb_games_db.csv", index_col = 0)

In [3]:
print("Total imdb_games: {} \nTotal giantbomb_games: {}".format(imdb_games.shape[0], giantbomb_games.shape[0]))

Total imdb_games: 14435 
Total giantbomb_games: 35117


We'll remove any exact duplicates from both dataframes

In [77]:
imdb_games = imdb_games.drop_duplicates()
giantbomb_games = giantbomb_games.drop_duplicates()
print("Total imdb_games: {} \nTotal giantbomb_games: {}".format(imdb_games.shape[0], giantbomb_games.shape[0]))

Total imdb_games: 14435 
Total giantbomb_games: 35116


Have a look at what our dataframes contain:

In [5]:
imdb_games.sample()

tconst  titleType primaryTitle originalTitle  isAdult startYear  \
1631306  tt11191330  videoGame      Asterix       Asterix        0      1983   

        endYear runtimeMinutes  genres  averageRating  numVotes         akas  
1631306      \N             \N  Action            5.5       7.0  ['Asterix']

In [6]:
giantbomb_games.sample()

franchise_id  game_id          name  release_year aliases  \
32102     3025-736     9228  Parasite Eve           NaN      PE   

                                  developers  \
32102  ['Squaresoft', 'Square Visual Works']   

                                        genres  \
32102  ['Action', 'Adventure', 'Role-Playing']   

                                               platforms  \
32102  ['PlayStation', 'PlayStation Network (PS3)', '...   

                        publishers       rating  
32102  ['Squaresoft', 'Square EA']  ['ESRB: M']

### Remove imdb duplicates

We want to remove any duplicates from our imdb data that contain the same name and year. We save the one with the highest number of votes

In [78]:
imdb_games = imdb_games.sort_values(by=["primaryTitle", "startYear", "numVotes"])
imdb_games = imdb_games.drop_duplicates(subset=["primaryTitle", "startYear"], keep="last")
imdb_games

tconst  titleType  \
7212109   tt4354918  videoGame   
5075818   tt1968978  videoGame   
367403    tt0383279  videoGame   
367404    tt0383280  videoGame   
4983813   tt1918633  videoGame   
...             ...        ...   
8014707   tt6169512  videoGame   
295707    tt0308989  videoGame   
770432    tt0795512  videoGame   
2185219   tt1219283  videoGame   
3830097  tt15235564  videoGame   

                                              primaryTitle  \
7212109                        #IDARB (It Draws a Red Box)   
5075818                                          '88 Games   
367403   'Goodbye, Galaxy!' Episode IV: Secret of the O...   
367404   'Goodbye, Galaxy!' Episode V: The Armageddon M...   
4983813                                           .detuned   
...                                                    ...   
8014707                                               osu!   
295707                                                 ssn   
770432                                            Æon Flux   
2185219                                 Îhatôvo monogatari   
3830097                                        ÜberSoldier   

                                             originalTitle  isAdult startYear  \
7212109                        #IDARB (It Draws a Red Box)        0      2015   
5075818                               Hyper Sports Special        0      1988   
367403   'Goodbye, Galaxy!' Episode IV: Secret of the O...        0      1991   
367404   'Goodbye, Galaxy!' Episode V: The Armageddon M...        0      1991   
4983813                                           .detuned        0      2009   
...                                                    ...      ...       ...   
8014707                                               osu!        0      2007   
295707                                                 ssn        0      1996   
770432                                            Æon Flux        0      2005   
2185219                                 Îhatôvo monogatari        0      1993   
3830097                                        ÜberSoldier        0      2006   

        endYear runtimeMinutes                   genres  averageRating  \
7212109      \N             \N                   Action            5.8   
5075818      \N             \N                    Sport            6.0   
367403       \N             \N  Action,Adventure,Sci-Fi            7.7   
367404       \N             \N  Action,Adventure,Sci-Fi            7.4   
4983813      \N             \N                    Music            2.3   
...         ...            ...                      ...            ...   
8014707      \N             \N            Music,Musical            7.4   
295707       \N             \N                       \N            7.9   
770432       \N             \N  Action,Adventure,Sci-Fi            5.6   
2185219      \N             \N                Adventure            7.0   
3830097      \N             \N  Action,Adventure,Sci-Fi            7.2   

         numVotes                                               akas  
7212109      25.0                    ['#IDARB (It Draws a Red Box)']  
5075818      14.0  ["Track & Field '88", 'Hyper Sports Special', ...  
367403      102.0  ["'Goodbye, Galaxy!' Episode IV: Secret of the...  
367404       82.0  ["'Goodbye, Galaxy!' Episode V: The Armageddon...  
4983813      28.0  ['.detuned', 'detuned: Gumi senpai no fushigi ...  
...           ...                                                ...  
8014707     107.0                                           ['osu!']  
295707       13.0                                            ['ssn']  
770432      265.0                                       ['Æon Flux']  
2185219       6.0            ['Ihatovo Story', 'Îhatôvo monogatari']  
3830097       7.0                                    ['ÜberSoldier']  

[14384 rows x 12 columns]

In [79]:
imdb_games.primaryTitle[imdb_games.duplicated(["primaryTitle", "startYear"], keep=False)].count()

0

### Remove giantbomb duplicates

As each game_id is unique, we want to group all the franchises for each game into a single result for each game_id. First though we check whether any of the the rows have the same game_id but a different name or release_year

In [80]:
giantbomb_games.count()

franchise_id    35116
game_id         35116
name            35116
release_year     5435
aliases          8088
developers      32746
genres          33660
platforms       34802
publishers      33604
rating          14184
dtype: int64

In [81]:
print(giantbomb_games.groupby("game_id").filter(lambda x: x["name"].nunique() > 1 or x["release_year"].nunique() > 1)["game_id"].count())

0


The next thing will be to remove any franchises which only contain 1 or 2 entries. Though they may technically count as a franchise, they do nothing to help us with analysis on the basis of longevity so we'll treat them as if they don't belong in a franchise

In [92]:
giantbomb_games.game_id.count()

35116

In [113]:
franchise_counts = giantbomb_games.groupby('franchise_id').size()
franchise_dict = {k:v for k,v in franchise_counts.items() if v < 3}
franchise_ids_to_drop = list(franchise_dict.keys())

mask = giantbomb_games['franchise_id'].isin(franchise_ids_to_drop)
giantbomb_games = giantbomb_games.drop(index=giantbomb_games[mask].index)

In [115]:
giantbomb_games.count()

franchise_id    31529
game_id         31529
name            31529
release_year     4668
aliases          7423
developers      29402
genres          30205
platforms       31251
publishers      30282
rating          13149
dtype: int64

The first thing to do is to combine all the games that have an identical franchise_id, release_year and name. With this criteria we will assume they are all the same game released on different platforms for example. We want to consider the release_year even when it is a nan value, so we convert all nan values to 0 accommodate this. Once we've done this, we group by game_id to combine the franchises for each individual game

In [117]:
giantbomb_games = giantbomb_games.groupby(["name", "franchise_id", giantbomb_games["release_year"].fillna(0)]).agg({
    "aliases": lambda x: list(set(x)),
    "developers": lambda x: list(set(x)),
    "genres": lambda x: list(set(x)),
    "platforms": lambda x: list(set(x)),
    "publishers": lambda x: list(set(x)),
    "rating": lambda x: list(set(x)),
    "game_id": "first"
}).reset_index().drop_duplicates(["game_id", "name", "franchise_id", "release_year"], keep="first")

giantbomb_games = giantbomb_games.groupby('game_id').agg({
    'franchise_id': lambda x: list(x),
    'name': 'first',
    'release_year': 'first',
    'aliases': 'first',
    'developers': 'first',
    'genres': 'first',
    "platforms": "first",
    "publishers" : "first",
    "rating": "first"
})

# Change year to string to match imdb database
giantbomb_games["release_year"] = giantbomb_games["release_year"].apply(lambda x : str(x).split('.')[0])

In [118]:
giantbomb_games.sort_values("game_id")

franchise_id                                name  \
game_id                                                                        
1                  [3025-143, 3025-2573]   Desert Strike: Return to the Gulf   
3                            [3025-2060]  Hyperballoid Deluxe: Survival Pack   
4                            [3025-1110]                The Chessmaster 2000   
6        [3025-128, 3025-130, 3025-1372]         WWE SmackDown! vs. RAW 2007   
8                             [3025-335]                    Super Spy Hunter   
...                                  ...                                 ...   
88822                        [3025-2291]                     Winning Post 10   
88824                        [3025-5696]              Wan Nyan Dōbutsu Byōin   
88831                        [3025-5699]        Shogi Saikyou: Pro ni Manabu   
88834                         [3025-383]    The Murder of Sonic the Hedgehog   
88837                        [3025-4309]                Morita Shogi Advance   

        release_year                  aliases  \
game_id                                         
1               1992  [Desert Strike Advance]   
3                  0                    [nan]   
4               1986                    [nan]   
6                  0               [SVR 2007]   
8                  0         [Battle Formula]   
...              ...                      ...   
88822              0                    [nan]   
88824              0                    [nan]   
88831              0                    [nan]   
88834              0                    [nan]   
88837              0                    [nan]   

                                                developers  \
game_id                                                      
1        [['Electronic Arts', 'Visual Concepts', 'Budca...   
3                                       [['Kernel Kaput']]   
4        [['The Software Toolworks', 'Software Country'...   
6        [["Yuke's Co. Ltd.", 'Digital Hearts Co., Ltd....   
8                                  [['Tokai Engineering']]   
...                                                    ...   
88822                                                [nan]   
88824                                                [nan]   
88831                                [['Magical Company']]   
88834                                           [['Sega']]   
88837                                                [nan]   

                                      genres  \
game_id                                        
1               [['Action', "Shoot 'Em Up"]]   
3             [['Action', 'Block-Breaking']]   
4        [['Strategy', 'Trivia/Board Game']]   
6        [['Action', 'Sports', 'Wrestling']]   
8                     [['Vehicular Combat']]   
...                                      ...   
88822                   [['Driving/Racing']]   
88824                       [['Simulation']]   
88831                [['Trivia/Board Game']]   
88834                        [['Adventure']]   
88837                [['Trivia/Board Game']]   

                                                 platforms  \
game_id                                                      
1        [['Amiga', 'Game Boy', 'Game Boy Advance', 'Ga...   
3                                                 [['PC']]   
4        [['Amiga', 'Amstrad CPC', 'Apple II', 'Atari S...   
6        [['PlayStation Portable', 'PlayStation 2', 'Xb...   
8                      [['Nintendo Entertainment System']]   
...                                                    ...   
88822    [['PC', 'PlayStation 4', 'Nintendo Switch', 'P...   
88824                               [['Game Boy Advance']]   
88831                                    [['PlayStation']]   
88834                                      [['Mac', 'PC']]   
88837                               [['Game Boy Advance']]   

                                                publishers  \
game_id                                         

We can now get a record of the number of duplicate names in the data

In [119]:
giantbomb_games.name[giantbomb_games.duplicated(["name"], keep=False)].count()

560

There is 594 duplicate names. Adding the criteria to distinguish by both name and release year returns the following:

In [120]:
giantbomb_games.name[giantbomb_games.duplicated(["name", "release_year"], keep=False)].count()

72

The release year needs to be converted to a string so we can compare it to the imdb database. We also need to tmove trailing decimal points.

In [121]:
giantbomb_games["release_year"] = giantbomb_games["release_year"].apply(lambda x : str(x).split('.')[0])

Unfortunately, at this stage, this is the furthest we can go for dealing with duplicates. The imdb datasets do not provide any additional information that can be corroborated across the two dataframes. As such there is no way to distinguish between a duplicate in the imdb dataset if the name and year are not unique in the giantbomb dataset. From this point, we now need to remove any duplicates in the imdb dataset, where there is not a clear distinction within the giantbomb dataset. The first thing I will do is isolate the duplicates. Then drop from the duplicates dataframe anywhere we can clearly identify which version of a game that duplicate is. Once we've finished we'll be left with a small dataframe of duplicates which we have tried all methods to identify the original game and failed. Every record left in the small dataframe will be dropped from our full imdb dataframe

Isolate the duplicates:

In [122]:
imdb_duplicates = imdb_games[imdb_games.duplicated(["primaryTitle"], keep=False)]

In [123]:
imdb_duplicates.tconst.count()

599

Remove any duplicate names we have a direct match for name and year.

In [124]:
matching_imdb_dups = pd.merge(imdb_games, giantbomb_games, left_on=['primaryTitle', 'startYear'], right_on=['name', 'release_year'], how='inner')
matching_imdb_dups

tconst  titleType                        primaryTitle  \
0    tt1968978  videoGame                           '88 Games   
1    tt0490765  videoGame                007: Licence to Kill   
2    tt0185834  videoGame                       10-Yard Fight   
3    tt1288449  videoGame  18 Wheels of Steel: Across America   
4    tt0185835  videoGame                                1942   
..         ...        ...                                 ...   
659  tt0896586  videoGame     Zoo Tycoon 2: African Adventure   
660  tt0910982  videoGame    Zoo Tycoon 2: Endangered Species   
661  tt0217905  videoGame                                Zool   
662  tt2976208  videoGame                              Zool 2   
663  tt1289679  videoGame                               Zorro   

                          originalTitle  isAdult startYear endYear  \
0                  Hyper Sports Special        0      1988      \N   
1                  007: Licence to Kill        0      1989      \N   
2                         10-Yard Fight        0      1983      \N   
3    18 Wheels of Steel: Across America        0      2003      \N   
4                                  1942        0      1984      \N   
..                                  ...      ...       ...     ...   
659     Zoo Tycoon 2: African Adventure        0      2006      \N   
660    Zoo Tycoon 2: Endangered Species        0      2005      \N   
661                                Zool        0      1992      \N   
662                              Zool 2        0      1993      \N   
663                               Zorro        0      1985      \N   

    runtimeMinutes                   genres_x  averageRating  ...  \
0               \N                      Sport            6.0  ...   
1               \N  Action,Adventure,Thriller            6.2  ...   
2               \N                      Sport            4.2  ...   
3               \N                     Family            6.8  ...   
4               \N                 Action,War            6.8  ...   
..             ...                        ...            ...  ...   
659             \N                         \N            7.0  ...   
660             \N                         \N            7.2  ...   
661             \N    Action,Adventure,Sci-Fi            6.6  ...   
662             \N           Action,Adventure            7.1  ...   
663             \N                         \N            7.8  ...   

                                                  akas franchise_id  \
0    ["Track & Field '88", 'Hyper Sports Special', ...  [3025-1053]   
1    ['Licence to Kill', 'Lizenz zum Töten', '007: ...   [3025-369]   
2    ['Ten Yado Faito', "10-Yard Fight '85", '10-Ya...  [3025-3081]   
3               ['18 Wheels of Steel: Across America']   [3025-747]   
4                        ['1942', 'Supercharger 1942']    [3025-11]   
..                                                 ...          ...   
659                ['Zoo Tycoon 2: African Adventure']   [3025-107]   
660               ['Zoo Tycoon 2: Endangered Species']   [3025-107]   
661  ['Zool', 'Zool: Ninja of the Nth Dimension', '...   [3025-768]   
662                                         ['Zool 2']   [3025-768]   
663                                          ['Zorro']  [3025-3239]   

                                   name release_year  \
0                             '88 Games         1988   
1                  007: Licence to Kill         1989   
2                         10-Yard Fight         1983   
3    18 Wheels of Steel: Across America         2003   
4                                  1942         1984   
..                                  ...          ...   
659     Zoo Tycoon 2: African Adventure         2006   
660    Zoo Tycoon 2: Endangered Species         2005   
661                                Zool         1992   
662                              Zool 2         1993   
663                               Zorro         1985   

                               

In [125]:
tempMask = ~imdb_duplicates['tconst'].isin(matching_imdb_dups['tconst'])
imdb_duplicates = imdb_duplicates[tempMask]
imdb_duplicates.tconst.count()

518

Remove any duplicates from our giantbomb database where we have no record of the year of release

In [170]:
gb_duplicate_names = giantbomb_games.duplicated(subset='name', keep=False)
release_year_int = giantbomb_games['release_year'].astype(int).eq(0)
gb_duplicates = giantbomb_games.loc[gb_duplicate_names & release_year_int]
#Having got all the duplicates with no release year, we'll drop them from the dataframe
giantbomb_games = giantbomb_games[~giantbomb_games.index.isin(gb_duplicates.index)]

At this stage, we are at an impasse with the imdb duplicates. We have no way to identify them from the information that we can cross-reference with the giantbomb data. If we attempt to match on the basis of name, then every duplicate entry in imdb will be considered to be the same game in the giantbomb dataset, which will be erroneous. If we arbritrarily remove one of the imdb duplicates, then the matched giantbomb entry may be for the wrong imdb game. As such, the only way we can try to prevent this from skewing the data is to completely drop any of those which we can't conclusively identify.

In [177]:
imdb_games = imdb_games[~imdb_games['tconst'].isin(imdb_duplicates['tconst'])]
imdb_games.tconst.count()

13866

This leaves me with 13866 games in the imdb dataset to play with. The next step is to combine the imdb dataset with the giantbomb

In [180]:
giantbomb_games.name.count()

25367

In [179]:
giantbomb_games.name.nunique()

25287

The way I am going to do this is to first combine the rows where the year and name matches. Then combine any remaining rows where the name matches but the year doesn't match.

In [205]:
copy_of_imdb_games = imdb_games
copy_of_giantbomb_games = giantbomb_games.reset_index()
print("Total imdb:{}, Total giantbomb:{}".format(copy_of_imdb_games.tconst.count(), copy_of_giantbomb_games.name.count()))

Total imdb:13866, Total giantbomb:25367


In [206]:
mergedDF = pd.merge(copy_of_imdb_games, copy_of_giantbomb_games, left_on=["primaryTitle", "startYear"], right_on=["name", "release_year"], how="inner")
copy_of_imdb_games = copy_of_imdb_games[~copy_of_imdb_games['tconst'].isin(mergedDF['tconst'])]
copy_of_giantbomb_games = copy_of_giantbomb_games[~copy_of_giantbomb_games['game_id'].isin(mergedDF['game_id'])]
print("Total imdb: {}, Total giantbomb: {}, Total in df: {}".format(copy_of_imdb_games.tconst.count(), copy_of_giantbomb_games.name.count(), mergedDF.tconst.count()))

Total imdb: 13202, Total giantbomb: 24703, Total in df: 664


In [207]:
copy_of_imdb_games.primaryTitle[copy_of_imdb_games.duplicated(["primaryTitle"], keep=False)].count()

0

In [220]:
copy_of_giantbomb_games[copy_of_giantbomb_games.duplicated(["name"], keep=False)].sort_values("release_year").head(30)

game_id franchise_id                 name release_year  \
13822    36496  [3025-4152]    Heavyweight Champ         1976   
10472    26738  [3025-4002]         Connect Four         1979   
11173    28623  [3025-3332]         Flash Gordon         1980   
15874    42189    [3025-25]                Alien         1981   
750       1962  [3025-2875]            King Kong         1982   
7903     20782   [3025-441]               Q*Bert         1982   
21690    67671  [3025-2875]            King Kong         1983   
1132      2957  [3025-3332]         Flash Gordon         1983   
23743    80398  [3025-3170]        Ashita no Joe         1983   
5961     15637   [3025-328]          Spy vs. Spy         1984   
21577    67114  [3025-1031]             Ultraman         1984   
1540      4049  [3025-1866]     The Wizard of Oz         1985   
24744    86278   [3025-184]          Family Feud         1985   
17212    46513   [3025-441]               Q*Bert         1986   
18601    51564   [3025-215]     Wheel of Fortune         1987   
21279    65421  [3025-4152]    Heavyweight Champ         1987   
10348    26424   [3025-184]          Family Feud         1987   
15239    40366   [3025-215]     Wheel of Fortune         1988   
376        988   [3025-375]            Jeopardy!         1988   
1133      2958   [3025-887]             Monopoly         1988   
5955     15621  [3025-2855]            Peter Pan         1988   
6095     16005   [3025-904]      The Flintstones         1988   
10978    28076  [3025-2587]          Judge Dredd         1990   
13397    35178  [3025-2875]            King Kong         1990   
23098    75985  [3025-3170]        Ashita no Joe         1990   
20754    62175   [3025-571]             Popeye 2         1990   
10345    26419   [3025-184]          Family Feud         1991   
14620    38715  [3025-4002]         Connect Four         1991   
14450    38255  [3025-1320]  Bo Jackson Baseball         1991   
7211     19094  [3025-2286]  Bo Jackson Baseball         1991   

                                                 aliases  \
13822                                              [nan]   
10472                                              [nan]   
11173                                              [nan]   
15874                                      [Super Alien]   
750                                                [nan]   
7903                                               [nan]   
21690                                              [nan]   
1132   [Spider City\r\nF18 (Vs) Aliens\r\nSpace Adven...   
23743                                              [nan]   
5961                                               [nan]   
21577                                              [nan]   
1540                                               [nan]   
24744                                              [nan]   
17212                                              [nan]   
18601                          [Glücksrad First Edition]   
21279                                              [nan]   
10348                                              [nan]   
15239                                              [nan]   
376                                                [nan]   
1133                                    [DX Monopoly GB]   
5955                                               [nan]   
6095                                               [nan]   
10978                                              [nan]   
13397                                              [nan]   
23098                                              [nan]   
20754                                              [nan]   
10345                                              [nan]   
14620                                        [Connect 4]   
14450                                              [nan]   
7211                                               [nan]   

                                              developers  \
13822                                         [['Sega']]   
10472     

In [213]:
mergedDF2 = pd.merge(copy_of_imdb_games, copy_of_giantbomb_games, left_on="primaryTitle", right_on="name", how="inner")
copy_of_imdb_games = copy_of_imdb_games[~copy_of_imdb_games['tconst'].isin(mergedDF2['tconst'])]
copy_of_giantbomb_games = copy_of_giantbomb_games[~copy_of_giantbomb_games['game_id'].isin(mergedDF2['game_id'])]
print("Total imdb: {}, Total giantbomb: {}, Total in df2: {}".format(copy_of_imdb_games.tconst.count(), copy_of_giantbomb_games.name.count(), mergedDF2.tconst.count()))

Total imdb: 7759, Total giantbomb: 19252, Total in df2: 5451


In [221]:
copy_of_giantbomb_games[copy_of_giantbomb_games.duplicated(["name"], keep=False)].sort_values("release_year").head(30)

game_id franchise_id                 name release_year  \
13822    36496  [3025-4152]    Heavyweight Champ         1976   
10472    26738  [3025-4002]         Connect Four         1979   
11173    28623  [3025-3332]         Flash Gordon         1980   
15874    42189    [3025-25]                Alien         1981   
750       1962  [3025-2875]            King Kong         1982   
7903     20782   [3025-441]               Q*Bert         1982   
21690    67671  [3025-2875]            King Kong         1983   
1132      2957  [3025-3332]         Flash Gordon         1983   
23743    80398  [3025-3170]        Ashita no Joe         1983   
5961     15637   [3025-328]          Spy vs. Spy         1984   
21577    67114  [3025-1031]             Ultraman         1984   
1540      4049  [3025-1866]     The Wizard of Oz         1985   
24744    86278   [3025-184]          Family Feud         1985   
17212    46513   [3025-441]               Q*Bert         1986   
18601    51564   [3025-215]     Wheel of Fortune         1987   
21279    65421  [3025-4152]    Heavyweight Champ         1987   
10348    26424   [3025-184]          Family Feud         1987   
15239    40366   [3025-215]     Wheel of Fortune         1988   
376        988   [3025-375]            Jeopardy!         1988   
1133      2958   [3025-887]             Monopoly         1988   
5955     15621  [3025-2855]            Peter Pan         1988   
6095     16005   [3025-904]      The Flintstones         1988   
10978    28076  [3025-2587]          Judge Dredd         1990   
13397    35178  [3025-2875]            King Kong         1990   
23098    75985  [3025-3170]        Ashita no Joe         1990   
20754    62175   [3025-571]             Popeye 2         1990   
10345    26419   [3025-184]          Family Feud         1991   
14620    38715  [3025-4002]         Connect Four         1991   
14450    38255  [3025-1320]  Bo Jackson Baseball         1991   
7211     19094  [3025-2286]  Bo Jackson Baseball         1991   

                                                 aliases  \
13822                                              [nan]   
10472                                              [nan]   
11173                                              [nan]   
15874                                      [Super Alien]   
750                                                [nan]   
7903                                               [nan]   
21690                                              [nan]   
1132   [Spider City\r\nF18 (Vs) Aliens\r\nSpace Adven...   
23743                                              [nan]   
5961                                               [nan]   
21577                                              [nan]   
1540                                               [nan]   
24744                                              [nan]   
17212                                              [nan]   
18601                          [Glücksrad First Edition]   
21279                                              [nan]   
10348                                              [nan]   
15239                                              [nan]   
376                                                [nan]   
1133                                    [DX Monopoly GB]   
5955                                               [nan]   
6095                                               [nan]   
10978                                              [nan]   
13397                                              [nan]   
23098                                              [nan]   
20754                                              [nan]   
10345                                              [nan]   
14620                                        [Connect 4]   
14450                                              [nan]   
7211                                               [nan]   

                                              developers  \
13822                                         [['Sega']]   
10472     

In [201]:
mergedDF3 = pd.merge(copy_of_imdb_games, copy_of_giantbomb_games, left_on="originalTitle", right_on="name", how="inner")
copy_of_imdb_games = copy_of_imdb_games[~copy_of_imdb_games['tconst'].isin(mergedDF3['tconst'])]
copy_of_giantbomb_games = copy_of_giantbomb_games[~copy_of_giantbomb_games['game_id'].isin(mergedDF3['game_id'])]
print("Total imdb: {}, Total giantbomb: {}, Total in df3: {}".format(copy_of_imdb_games.tconst.count(), copy_of_giantbomb_games.name.count(), mergedDF3.tconst.count()))

Total imdb: 7655, Total giantbomb: 19147, Total in df3: 105


In [204]:
mergedDF4 = pd.concat([mergedDF, mergedDF2, mergedDF3], axis=0)
mergedDF4

tconst  titleType                        primaryTitle  \
0    tt1968978  videoGame                           '88 Games   
1    tt0490765  videoGame                007: Licence to Kill   
2    tt0185834  videoGame                       10-Yard Fight   
3    tt1288449  videoGame  18 Wheels of Steel: Across America   
4    tt0185835  videoGame                                1942   
..         ...        ...                                 ...   
100  tt1235898  videoGame       Worldwide Soccer Manager 2008   
101  tt1325744  videoGame       Worldwide Soccer Manager 2009   
102  tt7451148  videoGame           Ys I: Ancient Ys Vanished   
103  tt0454989  videoGame         Ys VI: The Ark of Napishtim   
104  tt8751788  videoGame   Yu Yu Hakusho 100% Serious Battle   

                          originalTitle  isAdult startYear endYear  \
0                  Hyper Sports Special        0      1988      \N   
1                  007: Licence to Kill        0      1989      \N   
2                         10-Yard Fight        0      1983      \N   
3    18 Wheels of Steel: Across America        0      2003      \N   
4                                  1942        0      1984      \N   
..                                  ...      ...       ...     ...   
100               Football Manager 2008        0      2007      \N   
101               Football Manager 2009        0      2008      \N   
102              Ys: The Vanished Omens        0      1987      \N   
103            Ys: The Ark of Napishtim        0      2003      \N   
104      Yu Yu Hakusho 100% Maji Battle        0      2018      \N   

    runtimeMinutes                   genres_x  averageRating  ...  game_id  \
0               \N                      Sport            6.0  ...    43935   
1               \N  Action,Adventure,Thriller            6.2  ...     6264   
2               \N                      Sport            4.2  ...    10166   
3               \N                     Family            6.8  ...    14758   
4               \N                 Action,War            6.8  ...    10466   
..             ...                        ...            ...  ...      ...   
100             \N                      Sport            8.4  ...    20784   
101             \N                      Sport            8.3  ...    23498   
102             \N   Action,Adventure,Fantasy            7.2  ...    14318   
103             \N   Action,Adventure,Fantasy            7.9  ...    19621   
104             \N                     Action            5.3  ...    70198   

    franchise_id                                name release_year  \
0    [3025-1053]                           '88 Games         1988   
1     [3025-369]                007: Licence to Kill         1989   
2    [3025-3081]                       10-Yard Fight         1983   
3     [3025-747]  18 Wheels of Steel: Across America         2003   
4      [3025-11]                                1942         1984   
..           ...                                 ...          ...   
100   [3025-652]               Football Manager 2008            0   
101   [3025-652]               Football Manager 2009            0   
102   [3025-112]              Ys: The Vanished Omens            0   
103   [3025-112]            Ys: The Ark of Napishtim            0   
104   [3025-873]      Yu Yu Hakusho 100% Maji Battle            0   

                                               aliases  \
0                 [Konami '88\r\nHyper Sports Special]   
1    [Con Licencia para Matar\nPermis de Tuer\nLize...   
2                                                [nan]   
3                                                [nan]   
4           [Capcom Arcade Cabinet: 1942\n1942 MOBILE]   
..                                                 ...   
100  [Footy Manager 2008\nFootie Manager 2008\nFM08...   
101                    [Worldwide Soccer Manager 2009]   
102  [Ys: Ancient Ys Vanished\nYs 1\nAncient Land o...   
103                      [Ys VI: The Ark

### All this is just playing around. I'm using multiple dataframes to test it

The next step is to attempt to fuzzy match the titles and the original titles from the imdb dataset against the giantbomb database

In [26]:
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

I have 68 games with different originalTitles and primaryTitles, 

Check whether you have multiple entries with the same name and no year in GB. If it does, and 

In [174]:
imdb_duplicates

tconst  titleType             primaryTitle  \
5813970  tt2273075  videoGame                Action 52   
7523132  tt5058272  videoGame                Action 52   
8286665  tt6777472  videoGame  Adam's Venture: Origins   
8302190  tt6813690  videoGame  Adam's Venture: Origins   
8302192  tt6813694  videoGame  Adam's Venture: Origins   
...            ...        ...                      ...   
5475671  tt2132358  videoGame                    Zombi   
249553   tt0260590  videoGame               Zoo Keeper   
436868   tt0454991  videoGame               Zoo Keeper   
301459   tt0314957  videoGame               Zoo Tycoon   
7917082  tt5949676  videoGame               Zoo Tycoon   

                                             originalTitle  isAdult startYear  \
5813970                                          Action 52        0      1991   
7523132                                          Action 52        0      1993   
8286665  Adam's Venture: Episode 1, The Search for the ...        0      2009   
8302190        Adam's Venture: Episode 2, Solomon's Secret        0      2011   
8302192             Adam's Venture: Episode 3, Revelations        0      2012   
...                                                    ...      ...       ...   
5475671                                             ZombiU        0      2012   
249553                                          Zoo Keeper        0      1983   
436868                                          Zoo Keeper        0      2004   
301459                                          Zoo Tycoon        0      2001   
7917082                                         Zoo Tycoon        0      2013   

        endYear runtimeMinutes                   genres  averageRating  \
5813970      \N             \N    Action,Family,Fantasy            1.3   
7523132      \N             \N                       \N            3.1   
8286665      \N             \N                       \N            5.8   
8302190      \N             \N                       \N            7.5   
8302192      \N             \N                       \N            5.6   
...         ...            ...                      ...            ...   
5475671      \N             \N  Action,Adventure,Horror            6.3   
249553       \N             \N                   Family            7.0   
436868       \N             \N                   Action            6.1   
301459       \N             \N                   Action            7.4   
7917082      \N             \N                   Family            6.0   

         numVotes                                               akas  
5813970     141.0                                      ['Action 52']  
7523132      28.0                                      ['Action 52']  
8286665      15.0  ["Adam's Venture: Episode 1, The Search for th...  
8302190       9.0  ["Adam's Venture: Episode 2, Solomon's Secret"...  
8302192       9.0  ["Adam's Venture: Origins", "Adam's Venture: E...  
...           ...                                                ...  
5475671     259.0  ['ZombiU', 'Zombi', 'Killer Freaks from Outer ...  
249553       21.0                        ['Zoo Keeper', 'King Crab']  
436868       22.0                                     ['Zoo Keeper']  
301459      279.0                                     ['Zoo Tycoon']  
7917082      65.0                                     ['Zoo Tycoon']  

[518 rows x 12 columns]

In [139]:
temp = imdb_duplicates[~(imdb_duplicates["primaryTitle"]== imdb_duplicates["originalTitle"])]

In [134]:
temp = pd.merge(imdb_duplicates, giantbomb_games, left_on=['primaryTitle'], right_on=['name'], how='inner')
temp

tconst  titleType             primaryTitle  \
0    tt6777472  videoGame  Adam's Venture: Origins   
1    tt6813690  videoGame  Adam's Venture: Origins   
2    tt6813694  videoGame  Adam's Venture: Origins   
3    tt1978381  videoGame               Air Combat   
4    tt0429504  videoGame               Air Combat   
..         ...        ...                      ...   
624  tt4841246  videoGame      You Don't Know Jack   
625  tt0260590  videoGame               Zoo Keeper   
626  tt0454991  videoGame               Zoo Keeper   
627  tt0314957  videoGame               Zoo Tycoon   
628  tt5949676  videoGame               Zoo Tycoon   

                                         originalTitle  isAdult startYear  \
0    Adam's Venture: Episode 1, The Search for the ...        0      2009   
1          Adam's Venture: Episode 2, Solomon's Secret        0      2011   
2               Adam's Venture: Episode 3, Revelations        0      2012   
3                                           Air Combat        0      1993   
4                                           Ace Combat        0      1995   
..                                                 ...      ...       ...   
624                                You Don't Know Jack        0      2011   
625                                         Zoo Keeper        0      1983   
626                                         Zoo Keeper        0      2004   
627                                         Zoo Tycoon        0      2001   
628                                         Zoo Tycoon        0      2013   

    endYear runtimeMinutes                 genres_x  averageRating  ...  \
0        \N             \N                       \N            5.8  ...   
1        \N             \N                       \N            7.5  ...   
2        \N             \N                       \N            5.6  ...   
3        \N             \N                   Action            5.5  ...   
4        \N             \N  Action,Adventure,Sci-Fi            6.6  ...   
..      ...            ...                      ...            ...  ...   
624      \N             \N                       \N            6.4  ...   
625      \N             \N                   Family            7.0  ...   
626      \N             \N                   Action            6.1  ...   
627      \N             \N                   Action            7.4  ...   
628      \N             \N                   Family            6.0  ...   

                                                  akas franchise_id  \
0    ["Adam's Venture: Episode 1, The Search for th...  [3025-2362]   
1    ["Adam's Venture: Episode 2, Solomon's Secret"...  [3025-2362]   
2    ["Adam's Venture: Origins", "Adam's Venture: E...  [3025-2362]   
3                                       ['Air Combat']    [3025-16]   
4                         ['Ace Combat', 'Air Combat']    [3025-16]   
..                                                 ...          ...   
624                            ["You Don't Know Jack"]   [3025-115]   
625                        ['Zoo Keeper', 'King Crab']  [3025-4845]   
626                                     ['Zoo Keeper']  [3025-4845]   
627                                     ['Zoo Tycoon']   [3025-107]   
628                                     ['Zoo Tycoon']   [3025-107]   

                        name release_year  \
0    Adam's Venture: Origins            0   
1    Adam's Venture: Origins            0   
2    Adam's Venture: Origins            0   
3                 Air Combat         1992   
4                 Air Combat         1992   
..                       ...          ...   
624      You Don't Know Jack            0   
625               Zoo Keeper            0   
626               Zoo Keeper            0   
627               Zoo Tycoon            0   
628               Zoo Tycoon            0   

                                               aliases  \
0                                                [nan]   
1                      

In [74]:
giantbomb_games.sort_values("game_id").head()

franchise_id                                name  \
game_id                                                                        
1                  [3025-143, 3025-2573]   Desert Strike: Return to the Gulf   
3                            [3025-2060]  Hyperballoid Deluxe: Survival Pack   
4                            [3025-1110]                The Chessmaster 2000   
6        [3025-128, 3025-130, 3025-1372]         WWE SmackDown! vs. RAW 2007   
8                             [3025-335]                    Super Spy Hunter   

        release_year                  aliases  \
game_id                                         
1               1992  [Desert Strike Advance]   
3                  0                    [nan]   
4               1986                    [nan]   
6                  0               [SVR 2007]   
8                  0         [Battle Formula]   

                                                developers  \
game_id                                                      
1        [['Electronic Arts', 'Visual Concepts', 'Budca...   
3                                       [['Kernel Kaput']]   
4        [['The Software Toolworks', 'Software Country'...   
6        [["Yuke's Co. Ltd.", 'Digital Hearts Co., Ltd....   
8                                  [['Tokai Engineering']]   

                                      genres  \
game_id                                        
1               [['Action', "Shoot 'Em Up"]]   
3             [['Action', 'Block-Breaking']]   
4        [['Strategy', 'Trivia/Board Game']]   
6        [['Action', 'Sports', 'Wrestling']]   
8                     [['Vehicular Combat']]   

                                                 platforms  \
game_id                                                      
1        [['Amiga', 'Game Boy', 'Game Boy Advance', 'Ga...   
3                                                 [['PC']]   
4        [['Amiga', 'Amstrad CPC', 'Apple II', 'Atari S...   
6        [['PlayStation Portable', 'PlayStation 2', 'Xb...   
8                      [['Nintendo Entertainment System']]   

                                                publishers  \
game_id                                                      
1        [['Electronic Arts', 'Domark Software', 'Malib...   
3                         [['Alawar Entertainment, Inc.']]   
4        [['The Software Toolworks', 'Software Country'...   
6                                                [['THQ']]   
8                                            [['Sunsoft']]   

                                        rating  
game_id                                         
1                   [['ESRB: K-A', 'ESRB: E']]  
3                                        [nan]  
4                                        [nan]  
6        [['ESRB: T', 'PEGI: 16+', 'CERO: C']]  
8                                        [nan]

In [173]:
tempdb = pd.merge(imdb_duplicates, giantbomb_games, left_on=['primaryTitle'], right_on=['name'], how='inner')
tempdb

tconst  titleType             primaryTitle  \
0    tt6777472  videoGame  Adam's Venture: Origins   
1    tt6813690  videoGame  Adam's Venture: Origins   
2    tt6813694  videoGame  Adam's Venture: Origins   
3    tt1978381  videoGame               Air Combat   
4    tt0429504  videoGame               Air Combat   
..         ...        ...                      ...   
442  tt4841246  videoGame      You Don't Know Jack   
443  tt0260590  videoGame               Zoo Keeper   
444  tt0454991  videoGame               Zoo Keeper   
445  tt0314957  videoGame               Zoo Tycoon   
446  tt5949676  videoGame               Zoo Tycoon   

                                         originalTitle  isAdult startYear  \
0    Adam's Venture: Episode 1, The Search for the ...        0      2009   
1          Adam's Venture: Episode 2, Solomon's Secret        0      2011   
2               Adam's Venture: Episode 3, Revelations        0      2012   
3                                           Air Combat        0      1993   
4                                           Ace Combat        0      1995   
..                                                 ...      ...       ...   
442                                You Don't Know Jack        0      2011   
443                                         Zoo Keeper        0      1983   
444                                         Zoo Keeper        0      2004   
445                                         Zoo Tycoon        0      2001   
446                                         Zoo Tycoon        0      2013   

    endYear runtimeMinutes                 genres_x  averageRating  ...  \
0        \N             \N                       \N            5.8  ...   
1        \N             \N                       \N            7.5  ...   
2        \N             \N                       \N            5.6  ...   
3        \N             \N                   Action            5.5  ...   
4        \N             \N  Action,Adventure,Sci-Fi            6.6  ...   
..      ...            ...                      ...            ...  ...   
442      \N             \N                       \N            6.4  ...   
443      \N             \N                   Family            7.0  ...   
444      \N             \N                   Action            6.1  ...   
445      \N             \N                   Action            7.4  ...   
446      \N             \N                   Family            6.0  ...   

                                                  akas franchise_id  \
0    ["Adam's Venture: Episode 1, The Search for th...  [3025-2362]   
1    ["Adam's Venture: Episode 2, Solomon's Secret"...  [3025-2362]   
2    ["Adam's Venture: Origins", "Adam's Venture: E...  [3025-2362]   
3                                       ['Air Combat']    [3025-16]   
4                         ['Ace Combat', 'Air Combat']    [3025-16]   
..                                                 ...          ...   
442                            ["You Don't Know Jack"]   [3025-115]   
443                        ['Zoo Keeper', 'King Crab']  [3025-4845]   
444                                     ['Zoo Keeper']  [3025-4845]   
445                                     ['Zoo Tycoon']   [3025-107]   
446                                     ['Zoo Tycoon']   [3025-107]   

                        name release_year  \
0    Adam's Venture: Origins            0   
1    Adam's Venture: Origins            0   
2    Adam's Venture: Origins            0   
3                 Air Combat         1992   
4                 Air Combat         1992   
..                       ...          ...   
442      You Don't Know Jack            0   
443               Zoo Keeper            0   
444               Zoo Keeper            0   
445               Zoo Tycoon            0   
446               Zoo Tycoon            0   

                                               aliases  \
0                                                [nan]   
1                      

In [133]:
for index1 in imdb_duplicates.primaryTitle:
    #rint(index1)
    value = fuzz.token_sort_ratio("bction 2", index1)
    if value > 80:
        print(index1)
        print(value)

Action 52
82
Action 52
82


In [67]:
imdb_duplicates.head(500)

tconst  titleType             primaryTitle  \
5813970   tt2273075  videoGame                Action 52   
7523132   tt5058272  videoGame                Action 52   
8286665   tt6777472  videoGame  Adam's Venture: Origins   
8302190   tt6813690  videoGame  Adam's Venture: Origins   
8302192   tt6813694  videoGame  Adam's Venture: Origins   
...             ...        ...                      ...   
358721    tt0374336  videoGame            Wizard of Wor   
291841    tt0304947  videoGame           Wolfenstein 3D   
5558932  tt21632938  videoGame           Wolfenstein 3D   
8727036   tt7746896  videoGame                    Worms   
6646935   tt3059494  videoGame                    X-Men   

                                             originalTitle  isAdult startYear  \
5813970                                          Action 52        0      1991   
7523132                                          Action 52        0      1993   
8286665  Adam's Venture: Episode 1, The Search for the ...        0      2009   
8302190        Adam's Venture: Episode 2, Solomon's Secret        0      2011   
8302192             Adam's Venture: Episode 3, Revelations        0      2012   
...                                                    ...      ...       ...   
358721                                       Wizard of Wor        0      1981   
291841                                      Wolfenstein 3D        0      1992   
5558932                                     Wolfenstein 3D        0      1995   
8727036                                              Worms        0      2007   
6646935                                              X-Men        0      1992   

        endYear runtimeMinutes                   genres  averageRating  \
5813970      \N             \N    Action,Family,Fantasy            1.3   
7523132      \N             \N                       \N            3.1   
8286665      \N             \N                       \N            5.8   
8302190      \N             \N                       \N            7.5   
8302192      \N             \N                       \N            5.6   
...         ...            ...                      ...            ...   
358721       \N             \N    Action,Fantasy,Sci-Fi            7.7   
291841       \N             \N  Action,Adventure,Sci-Fi            8.0   
5558932      \N             \N        Action,Sci-Fi,War            6.7   
8727036      \N             \N                       \N            5.8   
6646935      \N             \N            Action,Sci-Fi            7.2   

         numVotes                                               akas  
5813970     141.0                                      ['Action 52']  
7523132      28.0                                      ['Action 52']  
8286665      15.0  ["Adam's Venture: Episode 1, The Search for th...  
8302190       9.0  ["Adam's Venture: Episode 2, Solomon's Secret"...  
8302192       9.0  ["Adam's Venture: Origins", "Adam's Venture: E...  
...           ...                                                ...  
358721       34.0                                  ['Wizard of Wor']  
291841     2242.0                                 ['Wolfenstein 3D']  
5558932       7.0                                 ['Wolfenstein 3D']  
8727036       7.0                                          ['Worms']  
6646935     110.0                                          ['X-Men']  

[500 rows x 12 columns]

In [57]:
for index1 in imdb_duplicates.primaryTitle:
    #rint(index1)
    value = fuzz.token_sort_ratio("bction 2", index1)
    if value > 80:
        print(index1)
        print(value)

Action 52
82
Action 52
82


In [ ]:
imdb_duplicates

In [ ]:
imdb_duplicates = imdb_duplicates[imdb_duplicates['primaryTitle'].isin(giantbomb_games['name'])]
imdb_duplicates

Create dictionary with count for each duplicate

In [ ]:
imdb_duplicate_counts

In [ ]:
imdb_duplicate_counts = imdb_duplicates.groupby('primaryTitle').size()
imdb_duplicate_dict = {k:v for k,v in imdb_duplicate_counts.items() if v > 1}
imdb_duplicate_dict

In [68]:
giantbomb_games[giantbomb_games["name"] == "Worms"]


franchise_id   name release_year aliases  \
game_id                                            
2007      [3025-133]  Worms         1995   [nan]   
20815     [3025-133]  Worms            0   [nan]   

                            developers                    genres  \
game_id                                                            
2007     [['Team17 Software Limited']]  [['Action', 'Strategy']]   
20815    [['Team17 Software Limited']]            [['Strategy']]   

                                                 platforms  \
game_id                                                      
2007     [['Amiga', 'Game Boy', 'Genesis', 'Super Ninte...   
20815    [['Xbox 360 Games Store', 'PlayStation Network...   

                                         publishers            rating  
game_id                                                                
2007     [['Telegames, Inc.', 'Limited Run Games']]             [nan]  
20815                 [['Team17 Software Limited']]  [['ESRB: E10+']]

Note here, we are only going to consider direct matches of a name for brevity's sake. If giantbomb contains no records that match the name of an imdb duplicate, we will assume the duplicates are not part of a franchise.

In [ ]:
for key, value in imdb_duplicate_dict.items():
    if giantbomb_games.name[giantbomb_games["name"] == key].count() == 0:
        # Game is not a franchise game, and can be kept in original imdb dataframe
        print(key)
        #imdb_duplicates = imdb_duplicates.drop(index=imdb_duplicates.index[imdb_duplicates['primaryTitle'] == key])
imdb_duplicates.count()

In [ ]:
imdb_duplicates.count()

In [ ]:
Action 52 is not in giantbomb games so we can keep in our original database as it's a unique game. Therefore we need to drop it from our dups dataframe

This leaves us with 411 duplicates, which exist in the imdb database but we may not be able to accurately match them with the 

WE ARE CREATING A DUPS DATAFRAME WHICH WILL BE ALL THE DUPLICATES WE CAN'T DISTINGUISH BETWENN. IF WE CAN DISTINGUISH, WE DROP FROM THE DUPS DATABASE. IF IMDB GAMES NAME DOES NOT APPEAR IN GIANTBOMB GAMES NAME, DROP FROM DUPLICATES LIST

sO FAR WE HAVE DONE ENOUGH. wE JUST NEED TO DO FUZZY MATCHING ON THE NAMES TO COMPARE.EG. ADAM'S VENTURE EXISTS IN BOTH BUT NEEDS FUZZY MATCHING AGAINST ORIGINAL TITLE

In [ ]:
imdb_duplicates = imdb_duplicates[~imdb_duplicates['tconst'].isin(matching_imdb_dups['tconst'])]


In [ ]:
for key, value in imdb_duplicate_dict.items():
    if giantbomb_games.name[giantbomb_games["name"] == key].count() == 0:
        # Game is not a franchise game, and can be kept in original imdb dataframe
        imdb_duplicates = imdb_duplicates.drop(index=imdb_duplicates.index[imdb_duplicates['primaryTitle'] == key])
    elif giantbomb_games.name[giantbomb_games["name"] == key].count() != value:
        #matching_ids holds index values of duplicates
        matching_imdb_ids=imdb_duplicates.index[imdb_duplicates['primaryTitle'] == key]
        matching_gb_ids = giantbomb_games.index[giantbomb_games['name'] == key]
        if 
        #for id in matching_imdb_ids:
            #This extracts the year
         #   year =imdb_duplicates.startYear[imdb_duplicates.index[imdb_duplicates.index == id]].item()
          #  if giantbomb_games.name[giantbomb_games["name"] == key] and giantbomb_games.release_year[giantbomb_games["release_year"] == year]:
           #     for 
            #    print("Great success!", year)
           # else:
            #    print("poo")
        
        print(key, value, giantbomb_games.name[giantbomb_games["name"] == key].count())

In [ ]:
giantbomb_games[giantbomb_games.duplicated(["name"], keep=False)].count()

In [ ]:
import re
temp = imdb_games[imdb_games["primaryTitle"].str.contains("Batman", na=False, flags=re.IGNORECASE, regex=True)]
temp

In [175]:
import re
temp2 = giantbomb_games[giantbomb_games["name"].str.contains("Adam's", na=False, flags=re.IGNORECASE, regex=True)]
temp2

franchise_id                                               name  \
game_id                                                                   
30427    [3025-2362]  Adam's Venture: Episode One - The Search for t...   
34686    [3025-2362]                Adam's Venture II: Solomon's Secret   
39166    [3025-2362]              Adam's Venture Episode 3: Revelations   
45160    [3025-2362]                          Adam's Venture Chronicles   
52907    [3025-2362]                            Adam's Venture: Origins   

        release_year          aliases           developers  \
game_id                                                      
30427           2009  [Adams Venture]  [['Vertigo Games']]   
34686           2011            [nan]  [['Vertigo Games']]   
39166              0            [nan]  [['Vertigo Games']]   
45160              0            [nan]  [['Vertigo Games']]   
52907              0            [nan]  [['Vertigo Games']]   

                                                genres  \
game_id                                                  
30427                                  [['Adventure']]   
34686    [['Adventure', 'Puzzle', 'Action-Adventure']]   
39166                                  [['Adventure']]   
45160                                  [['Adventure']]   
52907                                  [['Adventure']]   

                                                 platforms  \
game_id                                                      
30427                                             [['PC']]   
34686                                      [['Mac', 'PC']]   
39166                                             [['PC']]   
45160    [['PlayStation Network (PS3)', 'PC', 'PlayStat...   
52907    [['PC', 'Xbox One', 'PlayStation 4', 'Nintendo...   

                                 publishers                     rating  
game_id                                                                 
30427             [['Iceberg Interactive']]                      [nan]  
34686             [['Iceberg Interactive']]                      [nan]  
39166             [['Iceberg Interactive']]                      [nan]  
45160    [['Playlogic International N.V.']]  [['ESRB: E', 'PEGI: 7+']]  
52907                        [['Soedesco']]                      [nan]

In [ ]:
duplicate_giantbomb_names = giantbomb_games[giantbomb_games.duplicated(["name", "release_year"], keep=False)]
duplicate_giantbomb_names = duplicate_giantbomb_names.sort_values("name")
duplicate_giantbomb_names

Get list of all game_ids

In [ ]:
game_id_list = duplicate_giantbomb_names.index.to_list()

for each item in the list, get the corresponding json file. Clean to extract necessary data (esp. year)

In [ ]:
for game_id, api_url in game_api_dict.items():
    filename = "{}.json".format(game_id)
    if not Path(game_dir, filename).is_file():

        

In [ ]:
print(len(release_years))

In [ ]:
num_nans = id_year_df.isna().sum().sum()
print(num_nans)

In [ ]:
# group the data by name and count the number of occurrences
grouped = giantbomb_games.groupby(['name']).size().reset_index(name='count')

# select the rows where count > 1, indicating a duplicate name
duplicates = grouped[grouped['count'] > 1]

# select the rows with duplicate name and different game_id
duplicates = giantbomb_games[giantbomb_games['name'].isin(duplicates['name'])].groupby('name').filter(lambda x: x['game_id'].nunique() > 1)

# print the duplicate rows, sorted by name
if len(duplicates) > 0:
    print("Found {} rows with same name but different game_id:".format(len(duplicates)))
    print(duplicates.sort_values(by=['name', 'game_id']))
else:
    print("No rows with same name but different game_id found.")


In [ ]:
temp2 = temp[temp.duplicated(["name", "index"], keep=False)]
temp2 = temp2.sort_values("name")
temp2

In [ ]:
a = giantbomb_games[giantbomb_games["name"].str.contains('Sqrxz', na=False, flags=re.IGNORECASE, regex=True)]
a

The first thing I am going to do is check for any titles that appear only once in both dataframes and that have matching titles. With this criteria, we can assume that the games are the same. 

In [ ]:
unmatching_rows = imdb_games[~imdb_games['primaryTitle'].isin(giantbomb_games['name'])]

In [ ]:
try_this = unmatching_rows[unmatching_rows['originalTitle'].isin(giantbomb_games['name'])]
try_this.count()

In [ ]:
import re
temp = giantbomb_games[giantbomb_games["name"].str.contains('Infernal', na=False, flags=re.IGNORECASE, regex=True)]
temp.head(50)

My current plan for the most accurate way to approach this:

First check for duplicate names in both datasets. If there is a duplicate name in either dataset, we need to check whether this is because there are 2 games with the same name. If this is the case, we need a way to distinguish between the 2. I would anticipate year of release could work.

The next step would be comparing names against alternate titles. After this, any that still have no matches, would be subject to fuzzy matching.I'd anticipate at least 75% of the games in the imdb dataset should be contained in giantbomb's database.

In [ ]:
imdb_dups = imdb_games[imdb_games.duplicated(['primaryTitle', 'startYear'], keep=False)]
imdb_dups = imdb_dups.sort_values("primaryTitle")
print(imdb_dups)